scripts/smb-vuln-cve-2017-7494.nse

```diff
@@ -0,0 +1,511 @@
+
+          stdnse.debug1("Trying path : %s", curr_path)
+          status, result = smb.create_file(smbstate, curr_path, overrides)
+          stdnse.debug1("Status: %s, Result: %s", status, result)
+          -- on payload execution, result will be false and server will disconnect
+          if not status and string.match(result, "SMB: ERROR: Server disconnected the connection") then
+            samba_cve.state = vulns.STATE.VULN
+            table.insert(samba_cve.check_results,
+              "Exploitation of CVE-2017-7494 succeeded!")
+            return
+          end
+        end
+      end
+    end
+  end
+  if samba_cve.state ~= vulns.STATE.VULN and not path then
+    samba_cve.state = vulns.STATE.LIKELY_VULN
+    table.insert(samba_cve.check_results,
+      'File written to remote share, but unable to execute payload either due to unknown actual path, or the system may be patched.')
+  end
+end
+
+action = function(host,port)
+  local port = nmap.get_port_state(host,{number=smb.get_port(host),protocol='tcp'})
+
+  local result, stats
+  local response = {}
+
+  local samba_cve  = {
+    title = "SAMBA Remote Code Execution from Writable Share",
+    IDS = {CVE = 'CVE-2017-7494'},
+    risk_factor = "HIGH",
+    scores = {
+      CVSSv3 = "7.5 (HIGH) (CVSS:3.0/AV:N/AC:H/PR:L/UI:N/S:U/C:H/I:H/A:H)"
+    },
+    description = [[
+All versions of Samba from 3.5.0 onwards are vulnerable to a remote
+code execution vulnerability, allowing a malicious client to upload a
+shared library to a writable share, and then cause the server to load
+and execute it.
+]],
+    references = {
+      'https://www.samba.org/samba/security/CVE-2017-7494.html',
+    },
+    dates = {
+      disclosure = {year = '2017', month = '05', day = '24'},
+    },
+    check_results = {},
+    extra_info = {}
+  }
+
+  local report = vulns.Report:new(SCRIPT_NAME, host, port)
+  samba_cve.state = vulns.STATE.NOT_VULN
+
+  local check_version = stdnse.get_script_args(SCRIPT_NAME .. ".check-version") or false
+  -- check if they put false or similar
+  if check_version and string.lower(check_version) == "false" then
+    check_version = nil
+  end
+
+  local version = port.version.version
+
+  -- retrieve version of samba using smb.get_os
+  if not version then
+    local status, result = smb.get_os(host)
+
+    if(status == false) then
+      return stdnse.format_output(false, result)
+    end
+
+    -- result.lanmanager contains OS version information
+    -- string returned by result.lanmanager looks like Samba 4.3.9-Ubuntu
+    -- we only want 4.3.9-Ubuntu
+    if string.match(result.lanmanager,"^Samba ") then
+      version = string.match(result.lanmanager,"^Samba (.*)")
+    else
+      return stdnse.format_output(false,
+        "Either versioning failed or samba does not exist on the port!")
+    end
+  end
+
+  table.insert(samba_cve.check_results,
+    string.format("Samba Version: %s",version))
+
+  if check_version then
```

Comment:
We can make it complimentary, but I felt that the version numbers were not a strong enough heuristic to indicate vulnerability (due to the nature of the patches), and would have a lot of false positives. I have deliberately made it separate to support two different use cases: one to find which of the systems has yet to be updated on a network (via the version checks - this speeds up the scanning quite a bit) and the other to see if the system is exploitable, regardless of the version number.  

Do let me know if check_version should be complimentary and I'll make the necessary change!